# --- T2-COPYRIGHT-NOTE-BEGIN ---
# This copyright note is auto-generated by ./scripts/Create-CopyPatch.
# 
# T2 SDE: package/.../freeglut/gcc-4.0.patch
# Copyright (C) 2004 - 2005 The T2 SDE Project
# 
# More information can be found in the files COPYING and README.
# 
# This patch file is dual-licensed. It is available under the license the
# patched project is licensed under, as long as it is an OpenSource license
# as defined at http://www.opensource.org/ (e.g. BSD, X11) or under the terms
# of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# --- T2-COPYRIGHT-NOTE-END ---

Taken from Fedora Core to fix compilation with gcc-4.0.

diff -urN freeglut-2.2.0.orig/src/freeglut_callbacks.c freeglut-2.2.0/src/freeglut_callbacks.c
--- freeglut-2.2.0.orig/src/freeglut_callbacks.c	2003-12-11 02:32:09.000000000 +0100
+++ freeglut-2.2.0/src/freeglut_callbacks.c	2005-09-05 03:08:24.957884000 +0200
@@ -41,7 +41,7 @@
 #define SET_CALLBACK(a)              \
     if( fgStructure.Window == NULL ) \
         return;                      \
-    FETCH_WCB( ( *( fgStructure.Window ) ), a ) = callback;
+    SET_WCB( ( *( fgStructure.Window ) ), a, callback );
 
 /*
  * Sets the Display callback for the current window
diff -urN freeglut-2.2.0.orig/src/freeglut_internal.h freeglut-2.2.0/src/freeglut_internal.h
--- freeglut-2.2.0.orig/src/freeglut_internal.h	2003-12-11 18:02:13.000000000 +0100
+++ freeglut-2.2.0/src/freeglut_internal.h	2005-09-05 03:03:10.042203000 +0200
@@ -371,6 +371,30 @@
     GLboolean       NeedToResize;       /* Do we need to resize the window?  */
 };
 
+ /*
+ * SET_WCB() is used as:
+ *
+ *     SET_WCB( window, Visibility, func );
+ *
+ * ...where {window} is the freeglut window to set the callback,
+ *          {Visibility} is the window-specific callback to set,
+ *          {func} is a function-pointer.
+ *
+ * Originally, {FETCH_WCB( ... ) = func} was rather sloppily used,
+ * but this can cause warnings because the FETCH_WCB() macro type-
+ * casts its result, and a type-cast value shouldn't be an lvalue.
+ *
+ * The {if( FETCH_WCB( ... ) != func )} test is to do type-checking
+ * and for no other reason.  Since it's hidden in the macro, the
+ * ugliness is felt to be rather benign.
+ */
+#define SET_WCB(window,cbname,func)                   \
+do                                                    \
+{                                                     \
+    if( FETCH_WCB( window, cbname ) != func )         \
+        (((window).CallBacks[CB_ ## cbname]) = func); \
+} while( 0 )                                          \
+
 
 /*
  * FETCH_WCB() is used as:
diff -urN freeglut-2.2.0.orig/src/freeglut_structure.c freeglut-2.2.0/src/freeglut_structure.c
--- freeglut-2.2.0.orig/src/freeglut_structure.c	2003-12-11 02:32:09.000000000 +0100
+++ freeglut-2.2.0/src/freeglut_structure.c	2005-09-05 03:08:53.611674750 +0200
@@ -209,7 +209,7 @@
     {
         void *destroy = FETCH_WCB( *window, Destroy );
         fgClearCallBacks( window );
-        FETCH_WCB( *window, Destroy ) = destroy;
+        SET_WCB( *window, Destroy, destroy );
     }
 }
 
